larreco/RecoAlg/PMAlg/PmaSegment3D.cxx

```diff
@@ -0,0 +1,276 @@
+		double cosine = 0.0;
+		if (mag != 0.0) cosine = v0 * v1 / mag;
+		double b = v0Norm * cosine / v1Norm;
+
+		return vStart->Point3D() + (v3d * b);
+	}
+	else // segment 2D projection is almost a point
+	{
+		mf::LogWarning("pma::Segment3D") << "Short segment projection." << std::endl;
+
+		v3d = vStart->Point3D();
+		v3d += vStop->Point3D();
+		v3d *= 0.5;
+
+		return v3d;
+	}
+}
+
+void pma::Segment3D::SetProjection(pma::Hit3D& h) const
+{
+	pma::Node3D* vStart = static_cast< pma::Node3D* >(prev);
+	pma::Node3D* vStop = static_cast< pma::Node3D* >(next);
+
+	TVector2 v0(h.Point2D());
+	v0 -= vStart->Projection2D(h.View2D());
+
+	TVector2 v1(vStop->Projection2D(h.View2D()));
+	v1 -= vStart->Projection2D(h.View2D());
+
+	TVector3 v3d(vStop->Point3D());
+	v3d -= vStart->Point3D();
+
+	double v0Norm = v0.Mod();
+	double v1Norm = v1.Mod();
+
+	TVector2 p(vStart->Projection2D(h.View2D()));
+
+	double eps = 1.0E-6; // 0.01mm
+	if (v1Norm > eps)
+	{
+		double mag = v0Norm * v1Norm;
+		double cosine = 0.0;
+		if (mag != 0.0) cosine = v0 * v1 / mag;
+		double b = v0Norm * cosine / v1Norm;
+
+		p += (v1 * b);
+
+		h.SetProjection(p, (float)b);
+		h.SetPoint3D(vStart->Point3D() + (v3d * b));
+	}
+	else // segment 2D projection is almost a point
+	{
+		mf::LogWarning("pma::Segment3D") << "Short segment projection.";
+
+		p += vStop->Projection2D(h.View2D());
+		p *= 0.5F; h.SetProjection(p, 0.0F);
+
+		v3d = vStart->Point3D();
+		v3d += vStop->Point3D();
+		v3d *= 0.5;
+		h.SetPoint3D(v3d);
+	}
+}
+
+double pma::Segment3D::Length2(void) const
+{
+	if (prev && next)
+		return pma::Dist2( ((pma::Node3D*)prev)->Point3D(), ((pma::Node3D*)next)->Point3D() );
+	else
+	{
+		mf::LogError("pma::Segment3D") << "Segment endpoints not set.";
+		return 0.0;
+	}
+}
+
+
+double pma::Segment3D::GetDist2(const TVector3& psrc, const TVector3& p0, const TVector3& p1)
+{
+	TVector3 v0(psrc); v0 -= p0;
+	TVector3 v1(p1);   v1 -= p0;
+
+	TVector3 v2(psrc); v2 -= p1;
+	TVector3 v3(v1);   v3 *= -1.0;
+
+	double v0Norm2 = v0.Mag2();
+	double v1Norm2 = v1.Mag2();
+
+	double eps = 1.0E-6; // 0.01mm
+	if (v1Norm2 < eps)
+	{
+		mf::LogWarning("pma::Segment3D") << "Short segment or its projection.";
+		v1 = p0; v1 += p1; v1 *= 0.5;
+		return pma::Dist2(v1, psrc);
+	}
+
+	double mag01 = sqrt(v0Norm2 * v1Norm2);
+	double cosine01 = 0.0;
+	if (mag01 != 0.0) cosine01 = v0 * v1 / mag01;
+
+	double v2Norm2 = v2.Mag2();
+	double mag23 = sqrt(v2Norm2 * v3.Mag2());
+	double cosine23 = 0.0;
+	if (mag23 != 0.0) cosine23 = v2 * v3 / mag23;
+
+	double result = 0.0;
+	if ((cosine01 > 0.0) && (cosine23 > 0.0))
+	{
+		result = (1.0 - cosine01 * cosine01) * v0Norm2;
+	}
+	else // increase distance to prefer hit assigned to the vertex, not segment
+	{
+		if (cosine01 <= 0.0) result = 1.0001 * v0Norm2;
+		else result = 1.0001 * v2Norm2;
+	}
+
+	if (result >= 0.0) return result;
+	else return 0.0;
+}
+
+double pma::Segment3D::GetDist2(const TVector2& psrc, const TVector2& p0, const TVector2& p1)
+{
```

Comment:
Reminder about recommendation to find a replacement for TVector2.

The algorithm expressed in this function can be simplified by doing some more algebra by hand.